crates/gpui/src/platform/linux/wayland/clipboard.rs

```diff
@@ -0,0 +1,112 @@
+use std::io::{Read, Write};
+use std::os::fd::{AsRawFd, BorrowedFd, OwnedFd};
+
+use anyhow::{bail, Result};
+use filedescriptor::{FileDescriptor, Pipe};
+use wayland_client::protocol::wl_data_offer::{self, WlDataOffer};
+use wayland_client::protocol::wl_data_source::WlDataSource;
+
+use super::WaylandClient;
+use crate::ClipboardItem;
+
+pub const TEXT_MIME_TYPE: &str = "text/plain;charset=utf-8";
+
+pub(crate) struct Clipboard {
+    offer: Option<WlDataOffer>,
+    has_text: bool,
+    source: Option<WlDataSource>,
+    source_content: Option<String>,
+}
+
+impl Clipboard {
+    pub fn new() -> Self {
+        Self {
+            offer: None,
+            has_text: false,
+            source: None,
+            source_content: None,
+        }
+    }
+
+    pub fn receive_offer(&mut self, offer: WlDataOffer) {
+        self.offer = Some(offer);
+        self.has_text = false;
+    }
+
+    pub fn receive_mime_type(&mut self, mime: &str) {
+        self.has_text = self.has_text || mime == TEXT_MIME_TYPE;
+    }
+
+    pub fn send_source(&mut self, mime_type: &str, fd: OwnedFd) {
+        if let Some(ref mut source_content) = self.source_content {
+            let mut file = FileDescriptor::new(fd);
+            let _ = file.write(source_content.as_bytes());
+        }
```

Comment:
Ignoring the `file.write()` result, not sure if there would be a better behavior here.